### ISSUE TITLE
Explaining supported architectures in snap instructions

### ISSUE THREAD

**Issue by bmw** [certbot/website#619]
The Certbot snap which we plan to make our primary installation method supports the x86_64, ARMv7, and ARMv8 CPU architectures.

According to the [statistics](https://gist.github.com/bmw/7970ee3e3f4c8f8bad8e193a646c6c9b) I pulled from PyPI (which counts Certbot installations using `pip` including those from `certbot-auto`), this covers 97.8% of our user base. I also suspect this situation will get better over time as the older architectures continue to become less popular and supported.

The question I have is how should we approach this issue in Certbot's instruction generator? Should we make a nice glossary/tooltip item about this and explicitly list the supported architectures? I worry a bit about this bogging down the installation process and confusing novice users, most of whom won't have to worry about this, but that probably seems like the best approach to me personally.

**Comment by bmw** [comment]
For what it's worth, I pulled some new stats here which can be found at https://gist.github.com/bmw/c11817038a71f25609eb8f003f7c09e2.

**Comment by ohemorange** [comment]
Can snap itself be installed on the other architectures? The first step is to install `snapd`; if snapcraft's messages about why that doesn't work are sufficient, we may not need to worry about it.

**Comment by bmw** [comment]
>Can snap itself be installed on the other architectures?

Yes. For example, https://packages.debian.org/search?keywords=snapd&searchon=names&suite=stable&section=all shows Debian stable having `snapd` packaged for `amd64 arm64 armel armhf i386 ppc64el s390x`.

**Comment by bmw** [comment]
The error you get from the `snap install` command if you do install `snapd` though is:
```
# snap install certbot-dns-cloudflare
error: snap "certbot-dns-cloudflare" is not available on this architecture (i386) but exists on
       other architectures (amd64, arm64, armhf).
```
(The error is different on Certbot due to https://forum.snapcraft.io/t/is-it-possible-to-drop-an-unsupported-architecture-from-the-store/10510. That output is:
```
error: snap "certbot" is not available on stable but is available to install on the following
       channels:

       edge       snap install --edge certbot

       Please be mindful pre-release channels may include features not completely tested or
       implemented. Get more information with 'snap info certbot'.
```
)

**Comment by bmw** [comment]
To be clear, I don't have a strong opinion on whether or not we have a note. With that said, to state my preference, I suppose I slightly prefer saying something but at least in the long term documenting it very well like we do all of our other confusing terminology.

Maybe adding it here in the generated snap instructions makes the most sense?
![Screen Shot 2020-07-28 at 3 11 55 PM](https://user-images.githubusercontent.com/6504915/88727555-b241d080-d0e4-11ea-8f49-119331441209.png)

**Comment by ohemorange** [comment]
I still haven't found a solution I love. I think it'll be missed by people migrating if we put it in the icons/prerequisites section, plus, that's a shared section, so we'd have to add a snap-specific version. Maybe add a step before "Install snapd" like "Confirm your architecture is supported: The Certbot snap only supports the x86_64, ARMv7, and ARMv8 architectures (tooltip with definition). To see your architecture, try running `uname -m`."

**Comment by bmw** [comment]
That'd work for me.

The concern that comes to mind is the strings used to designate these architectures varies quite a bit and IIRC the output of `uname -m` isn't standardized and there isn't a standarized option that is widely available. See https://github.com/certbot/certbot/pull/7587#discussion_r353473982. The output of `uname -m` is probably pretty consistent on Linux in practice though, but I think we'd probably want to make sure our instructions include the output they should expect if their architecture is supported.

This is a slight tangent, but another issue here that initially worried me quite a bit is what we should say, if anything, if the user's architecture isn't supported by our snaps. At least for now, if we want to include text about this, I personally think we should just say other architectures are not supported by Certbot. We could always reevaluate this if a significant demand arises, but I think doing anything else complicates things quite a bit.

**Comment by ohemorange** [comment]
> The concern that comes to mind is the strings used to designate these architectures varies quite a bit and IIRC the output of uname -m isn't standardized and there isn't a standarized option that is widely available. See certbot/certbot#7587 (comment). The output of uname -m is probably pretty consistent on Linux in practice though, but I think we'd probably want to make sure our instructions include the output they should expect if their architecture is supported.

Of course I should know better than to expect it would be that easy. I do think since it's human-readable output we're looking for, that should be sufficiently manageable though. I definitely like the idea of including expected outputs.

> This is a slight tangent, but another issue here that initially worried me quite a bit is what we should say, if anything, if the user's architecture isn't supported by our snaps. At least for now, if we want to include text about this, I personally think we should just say other architectures are not supported by Certbot. We could always reevaluate this if a significant demand arises, but I think doing anything else complicates things quite a bit.

Great point, I agree. "Try a different installation method"? "Selecting a different system from the dropdown"?

**Comment by bmw** [comment]
> "Try a different installation method"? "Selecting a different system from the dropdown"?

I was proposing we don't suggest anything like this right now.

My plan was to start slowly transitioning the instructions for different OSes to our snap instructions. If we do that for say our Ubuntu instructions and you're an Ubuntu user on another architecture, where should we direct you? The change I made removed our instructions using OS packages, `certbot-auto` is going away, and our Docker images support the same architectures as the snaps. Even worse, what do our "Other UNIX" instructions say if they're updated to direct people towards `snapd` and not certbot-auto?

**Comment by ohemorange** [comment]
......ok, I'm now officially for "leaving it out for now". I also think it's fine to direct people back to the ubuntu page as long as it's around; "directing people" and "setting up an automatic redirect" are two different levels. In the long run, we do have developer instructions for how to get things running from source, which could either be the official instructions or a basis for them. But I think the right approach is:

short term: do nothing
medium term, if it's a problem: just point back to OS pages, that should get many people I think?
long term: install from source/pip (we are still going to be releasing to pip, right?)

do we know what OSes the people running on weird architectures tend to be using? like, it's all, say, raspbian, that would make this problem easier to address.

**Comment by bmw** [comment]
> I also think it's fine to direct people back to the ubuntu page as long as it's around
> ...
> medium term, if it's a problem: just point back to OS pages, that should get many people I think?

Well my plan was to do things like having the "Ubuntu" dropdown option generate our snap instructions so the instructions using Ubuntu OS packages wouldn't be around anymore. We could certainly revisit this if it becomes a problem though. My concern with keeping it around is I think it will result in more users using those old packages rather than the snap.

> do we know what OSes the people running on weird architectures tend to be using?

Good question. Keeping in mind that [the Debian 8 numbers seem unrealistically inflated](https://github.com/certbot/website/issues/619#issuecomment-665196794), I pulled stats on this at https://gist.github.com/518188574c700883e3a14a08c620d668.

**Comment by bmw** [comment]
> we are still going to be releasing to pip, right?

Whoops. Missed this one.

Yes. I personally think we should keep releasing to PyPI allowing Certbot to be installed through tools like `pip`. That's where our OS packagers pull from and while I'm certainly not against considering it if we see a benefit, I'm not aware of any significant open source Python packages in the world that aren't released on PyPI.

**Comment by ohemorange** [comment]
> Well my plan was to do things like having the "Ubuntu" dropdown option generate our snap instructions so the instructions using Ubuntu OS packages wouldn't be around anymore. We could certainly revisit this if it becomes a problem though. My concern with keeping it around is I think it will result in more users using those old packages rather than the snap.

Hm, that's fair. We could even keep the generated page around and link directly to it, and still have the dropdown go to snaps.

> Keeping in mind that the Debian 8 numbers seem unrealistically inflated, I pulled stats on this at https://gist.github.com/518188574c700883e3a14a08c620d668.

That's unfortunate, I was hoping it would correspond mostly to rarer OSes.

> Yes. I personally think we should keep releasing to PyPI allowing Certbot to be installed through tools like pip. That's where our OS packagers pull from and while I'm certainly not against considering it if we see a benefit, I'm not aware of any significant open source Python packages in the world that aren't released on PyPI.

That should make developing backup instructions even easier.

**Comment by bmw** [comment]
> Hm, that's fair. We could even keep the generated page around and link directly to it, and still have the dropdown go to snaps.

Good idea! I think that'd take some tinkering to continue to generate those instructions when no dropdown option creates them, but I think we could make that work.

It sounds like we're on the same page for now though and I think we can revisit this if it becomes a problem.

**Comment by ohemorange** [comment]
> I think that'd take some tinkering to continue to generate those instructions when no dropdown option creates them, but I think we could make that work.

Agreed, but I don't think it would be too bad.

> It sounds like we're on the same page for now though and I think we can revisit this if it becomes a problem.

Agreed! Glad we have a plan.

**Comment by bmw** [comment]
@ohemorange, do you know how the tooltip system works? Does it mainly just require us writing the text we want or is some web dev work involved?

I'm trying to decide if it makes sense for us to ask E&D for help on this.

**Comment by ohemorange** [comment]
Yes, I do, it's very easy. You add a definition to the glossary, and use a tag instead of the text where you want the tooltip to be.
